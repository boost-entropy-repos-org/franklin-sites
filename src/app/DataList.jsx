import React from 'react';
import PropTypes from 'prop-types';
import Card from '../components/card';
import ScrollDataLoader from '../components/scroll-data-loader';

const DataList = ({
  data, loadMoreData, hasMoreData, idKey, onSelect, selected,
}) => (
  <div className="data-list">
    <div className="data-list__wrapper">
      <ScrollDataLoader
        idKey={idKey}
        onLoadMoreData={loadMoreData}
        hasMoreData={hasMoreData}
        items={data.map(({ [idKey]: id, content }) => (
          <Card key={id} selectable selected={!!selected[id]} onSelect={() => onSelect(id)}>
            <p>{content}</p>
          </Card>
        ))}
      />
    </div>
  </div>
);

DataList.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  loadMoreData: PropTypes.func.isRequired,
  hasMoreData: PropTypes.bool.isRequired,
  idKey: PropTypes.string,
  onSelect: PropTypes.func.isRequired,
  selected: PropTypes.instanceOf(Object).isRequired,
};

DataList.defaultProps = {
  idKey: 'id',
};

export default DataList;
