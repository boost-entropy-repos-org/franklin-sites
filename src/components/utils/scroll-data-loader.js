const scrollOffsetPercentage = 10;
const scrollOffsetFactor = 1 + scrollOffsetPercentage / 100;

export const isNotScrollable = (node) => {
  const { scrollHeight, offsetHeight } = node.current;
  return scrollHeight <= offsetHeight * scrollOffsetFactor;
};

export const isBottom = (node) => {
  const { scrollHeight, scrollTop, offsetHeight } = node.current;
  return scrollHeight - Math.ceil(scrollTop) <= offsetHeight * scrollOffsetFactor;
};

export const checkLoadMoreItems = (loading, hasMoreData, ref, setLoadMoreItems) => {
  if (!loading && hasMoreData && isBottom(ref)) {
    setLoadMoreItems(true);
  }
};

export const checkOnDataLoad = (
  hasMoreData,
  setLoading,
  onLoadMoreItems,
  setLoadMoreItems,
  ref,
) => {
  if (hasMoreData && isNotScrollable(ref)) {
    setLoading(true);
    onLoadMoreItems();
  } else {
    setLoading(false);
    setLoadMoreItems(false);
  }
};

export const checkOnLoadMoreItems = (loadMoreItems, setLoading, onLoadMoreItems) => {
  if (loadMoreItems) {
    setLoading(true);
    onLoadMoreItems();
  }
};
