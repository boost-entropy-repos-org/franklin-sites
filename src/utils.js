export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, value, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, value, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!value || thisNode.value === value) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

function getSquashedPath(items, sep = ' / ') {
  return {
    value: items[items.length - 1].value,
    pathLabel: items.map(item => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
    items,

  };
}

export function getSquashedArrayOfPaths(paths) {
  // console.log(paths);
  return paths.map(items => getSquashedPath(items));
}
