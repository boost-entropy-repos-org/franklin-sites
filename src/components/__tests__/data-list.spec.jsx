import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DataList from '../data-list';
import { fillArray } from '../../utils';

describe('DataList', () => {
  const data = fillArray(10, (element, index) => ({
    id: `id${index}`,
    content: index,
  }));

  let onLoadMoreItems;
  let dataRenderer;
  beforeEach(() => {
    onLoadMoreItems = jest.fn();
    dataRenderer = item => <p>{item.content}</p>;
  });

  test('should render', () => {
    const { asFragment } = render(
      <DataList
        onLoadMoreItems={onLoadMoreItems}
        hasMoreData
        data={data}
        dataRenderer={dataRenderer}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should request more data', () => {
    const { container } = render(
      <DataList
        onLoadMoreItems={onLoadMoreItems}
        hasMoreData
        data={data}
        dataRenderer={dataRenderer}
        onCardClick={null}
      />
    );
    const scrollContainer = container.querySelector(
      '.data-loader__scroll-container'
    );
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  test('should not request more data', () => {
    const { container } = render(
      <DataList
        onLoadMoreItems={onLoadMoreItems}
        hasMoreData={false}
        data={data}
        dataRenderer={dataRenderer}
        onCardClick={null}
      />
    );
    const scrollContainer = container.querySelector(
      '.data-loader__scroll-container'
    );
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalledTimes(0);
  });
});
