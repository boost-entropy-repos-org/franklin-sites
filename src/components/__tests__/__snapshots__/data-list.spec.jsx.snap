// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList should render 1`] = `
<DocumentFragment>
  <div
    class="data-loader__wrapper"
  >
    <div
      class="data-loader__scroll-container"
    >
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            0
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            1
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            2
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            3
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            4
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            5
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            6
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            7
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            8
          </p>
        </div>
      </div>
      <div
        class="card  "
      >
        <div
          class="card__content"
        >
          <p>
            9
          </p>
        </div>
      </div>
      <div
        class="data-loader__loading"
      >
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
